ethereum/p2p/src/test/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshStateTest.java

```diff
@@ -0,0 +1,333 @@
+    verify(neighborFinder).issueFindNodeRequest(peer_000);
+
+    recursivePeerRefreshState.digestNeighboursPacket(neighborsPacketData_000, peer_000);
+
+    verify(neighborFinder, never()).issueFindNodeRequest(peer_010);
+    verify(neighborFinder).issueFindNodeRequest(peer_011);
+    verify(neighborFinder).issueFindNodeRequest(peer_012);
+    verify(neighborFinder).issueFindNodeRequest(peer_013);
+
+    TimeUnit.SECONDS.sleep(60);
+
+    verify(neighborFinder).issueFindNodeRequest(peer_010);
+  }
+
+  private TestPeer generatePeer(final JsonNode peer) {
+    int parent = peer.get("parent").asInt();
+    int tier = peer.get("tier").asInt();
+    int identifier = peer.get("identifier").asInt();
+    int ordinalRank = peer.get("ordinalRank").asInt();
+    BytesValue id = BytesValue.fromHexString(peer.get("id").asText());
+    List<Peer> peerTable = new ArrayList<>();
+    if (peer.get("peerTable") != null) {
+      JsonNode peers = peer.get("peerTable");
+      for (JsonNode element : peers) {
+        peerTable.add(generatePeer(element));
+      }
+    } else {
+      peerTable = Collections.emptyList();
+    }
+    return new TestPeer(parent, tier, identifier, ordinalRank, id, peerTable);
+  }
+
+  private boolean matchPeerToCorrespondingPacketData(
+      final Peer peer, final NeighborsPacketData neighborsPacketData) {
+    String nodeId = null;
+    Matcher idMatcher = Pattern.compile("\\d\\.\\d\\.(\\d)").matcher(peer.toString());
+    if (idMatcher.find()) {
+      nodeId = idMatcher.group(1);
+    }
+    for (Peer neighbour : neighborsPacketData.getNodes()) {
+      idMatcher = Pattern.compile("(\\d)\\.(\\d)\\.(\\d)").matcher(neighbour.toString());
+      if (idMatcher.find()) {
+        String parentId = idMatcher.group(1);
+        if (!nodeId.equals(parentId)) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  static class TestPeer implements Peer {
+    int parent;
+    int tier;
+    int identifier;
+    int ordinalRank;
+    BytesValue id;
+    List<Peer> peerTable;
+
+    TestPeer(
+        final int parent,
+        final int tier,
+        final int identifier,
+        final int ordinalRank,
+        final BytesValue id,
+        final List<Peer> peerTable) {
+      this.parent = parent;
+      this.tier = tier;
+      this.identifier = identifier;
+      this.ordinalRank = ordinalRank;
+      this.id = id;
+      this.peerTable = peerTable;
+    }
+
+    int getOrdinalRank() {
+      return ordinalRank;
+    }
+
+    List<Peer> getPeerTable() {
+      return peerTable;
+    }
+
+    @Override
+    public BytesValue getId() {
+      return this.id;
+    }
+
+    @Override
+    public Bytes32 keccak256() {
+      return null;
+    }
+
+    @Override
+    public Endpoint getEndpoint() {
+      return null;
+    }
+
+    @Override
+    public String toString() {
+      return parent + "." + tier + "." + identifier;
+    }
+  }
+
+  @After
+  public void cleaUp() {
+    vertx.close();
+  }
```

Comment:
I'd move this up to under init so that you can see more clearly that we start a Vertx instance and close it rather than having the two so separated.  Although with the timer becoming external you'll get rid of the need for Vertx in this test entirely which is another nice benefit.